### PR TITLE
feat: implement groth16

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [groth16 proving scheme](https://eprint.iacr.org/2016/260.pdf). Please refer to https://github.com/arkworks-rs/groth16, too!

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/zk/r1cs/constraint_system/quadratic_arithmetic_program.h [hunk 1]
```diff
@@ -91,6 +99,92 @@ class QuadraticArithmeticProgram {
...
+
+    Evals& ab_evals = a_evals *= b_evals;
+
+    F vanishing_polynomial_over_coset =
+        domain
+            ->EvaluateVanishingPolynomial(
+                F::FromMontgomery(F::Config::kSubgroupGenerator))
+            .Inverse();
+
+    // |h_evals[i]| = (|a[i]| * |b[i]| - c[i)) / (g * ωⁿ⁺ˡ⁺¹ - 1)
```
Comment (Insun35) [inline tachyon/zk/r1cs/constraint_system/quadratic_arithmetic_program.h:177]:
```suggestion
    // |h_evals[i]| = (|a[i]| * |b[i]| - |c[i]|)) / (g * ωⁿ⁺ˡ⁺¹ - 1)
```

File: tachyon/zk/r1cs/groth16/prove.h [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  using G1Bucket = typename math::VariableBaseMSM<G1AffinePoint>::Bucket;
+  using G2Bucket = typename math::VariableBaseMSM<G2AffinePoint>::Bucket;
+
+  math::VariableBaseMSM<G1AffinePoint> msm;
+
+  // |witness_acc| = [Σᵢ₌ₗ₊₁..ₘ (β * aᵢ(x) + α * bᵢ(x) + cᵢ(x)) / δ]₁
+  G1Bucket witness_acc;
+  CHECK(msm.Run(pk.l_g1_query(), witness_assignments, &witness_acc));
+
+  // |h_acc| = |(h(x) * t(x)) / δ]₁
```
Comment (Insun35) [inline tachyon/zk/r1cs/groth16/prove.h:57]:
```suggestion
  // |h_acc| = [(h(x) * t(x)) / δ]₁

**Review by ashjeong — COMMENTED** [review]
https://github.com/kroma-network/tachyon/pull/329/commits/a26fc6050bed195384d6db4387723521bb9f27d8
“Add indexing operator on Matrix” -> “Add indexing operator to Matrix”

https://github.com/kroma-network/tachyon/pull/329/commits/59aea9437dfe610074c1f9f685204e4ea39a7187
“See
some parts of…” -> “See some parts of…”

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**